File: src/UE4_Localization/FTextLocalizationResource.java
/**
 * 
 */
package UE4_Localization;

import java.util.ArrayList;
import java.util.List;

import UE4.FArchive;
import UE4.deserialize.exception.DeserializationException;
import UE4_Assets.FGUID;
import UE4_Assets.ReadException;
import annotation.CustomSerializable;
import lombok.Data;

/**
 * @author FunGames
 *
 */
@Data
@CustomSerializable
public class FTextLocalizationResource {
	private byte version;
	private List<LocaleNamespace> stringData;
	private long strArrayOffset;

	public static final FGUID LOCRES_MAGIC = new FGUID("1970541582-4228074087-2643465546-461322179");
	private static final int INDEX_NONE = -1;

	@SuppressWarnings("unused")
	public FTextLocalizationResource(FArchive Ar) throws ReadException, DeserializationException {
		FGUID magic = Ar.read(FGUID.class);

		if (!magic.getString().equals(LOCRES_MAGIC.getString())) {
			throw new ReadException("Wrong locres guid");
		}

		version = Ar.readUInt8();
		strArrayOffset = Ar.readInt64();
		if (strArrayOffset == INDEX_NONE) {
			throw new ReadException("No offset found");
		}

		// Only works for version: optimized
		int currentOffset = Ar.Tell();

		Ar.Seek((int) strArrayOffset);
		List<FTextLocalizationResourceString> localizedStrings = new ArrayList<>();
		long localizedStringCount = Ar.readUInt32();
		for (int i = 0; i < localizedStringCount; i++) {
			localizedStrings.add(Ar.read(FTextLocalizationResourceString.class));
		}

		Ar.Seek(currentOffset);

		long entryCount = Ar.readUInt32();
		long namespaceCount = Ar.readUInt32();
		this.stringData = new ArrayList<>();
		for (int i = 0; i < namespaceCount; i++) {
			FTextKey namespace = Ar.read(FTextKey.class);
			long keyCount = Ar.readUInt32();

			List<FEntry> strings = new ArrayList<>();
			
			for(int j = 0; j < keyCount; j++) {
				FTextKey textKey = Ar.read(FTextKey.class);
				long _sourceHash = Ar.readUInt32();
				int stringIndex = Ar.readInt32();
				if(stringIndex > 0 && stringIndex < localizedStrings.size()) {
					strings.add(new FEntry(textKey.getText(), localizedStrings.get(stringIndex).getData()));
				}
			}
			
			stringData.add(new LocaleNamespace(namespace.getText(), strings));
		}
	}
}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/locres/objects/FTextLocalizationResource.kt
package me.fungames.jfortniteparse.ue4.locres.objects

import me.fungames.jfortniteparse.exceptions.ParserException
import me.fungames.jfortniteparse.ue4.FGuid
import me.fungames.jfortniteparse.ue4.UClass
import me.fungames.jfortniteparse.ue4.reader.FArchive

@ExperimentalUnsignedTypes
class FTextLocalizationResource : UClass {
    companion object {
        val locresMagic = FGuid(1970541582u, 4228074087u, 2643465546u, 461322179u)
        const val indexNone = -1L
    }

    var version : UByte
    var strArrayOffset: Long
    val stringData : MutableMap<String, MutableMap<String, String>>

    constructor(Ar : FArchive) {
        super.init(Ar)
        val magic = FGuid(Ar)
        if (magic != locresMagic)
            throw ParserException("Wrong locres guid")
        version = Ar.readUInt8()
        strArrayOffset = Ar.readInt64()
        if (strArrayOffset == indexNone)
            throw ParserException("No offset found")

        //Only works for version 'optimized'
        val cOffset = Ar.pos()

        Ar.seek(strArrayOffset.toInt())
        val localizedStrings = Ar.readTArray { FTextLocalizationResourceString(Ar) }
        Ar.seek(cOffset)

        Ar.readUInt32() // entryCount
        val nameSpaceCount = Ar.readUInt32()
        stringData = mutableMapOf()
        for (i in 0 until nameSpaceCount.toInt()) {
            val nameSpace = FTextKey(Ar)
            val keyCount = Ar.readUInt32()

            val strings = mutableMapOf<String, String>()
            for (j in 0 until keyCount.toInt()) {
                val textKey = FTextKey(Ar)
                Ar.readUInt32() // source hash
                val stringIndex = Ar.readInt32()
                if (stringIndex > 0 && stringIndex < localizedStrings.size)
                    strings[textKey.text] = localizedStrings[stringIndex].data
            }
            stringData[nameSpace.text] = strings
        }
        super.complete(Ar)
    }
}
